File: sdks/python/perf_harness/drive/scheduler.py
"""One bounded scheduler for finite arrivals and concurrency-driven replenishment."""

from __future__ import annotations

import asyncio
import random
import time
from contextlib import suppress
from dataclasses import dataclass, field, replace

from harness_common import Operation, OperationRun
from spec_case.model import Case

from perf_harness.drive.runner import ArmContext, FireContext, Runner
from perf_harness.judge import Judge
from perf_harness.model import ArmRun, ArmStop, Outcome, StopSnapshot
from perf_harness.observe.base import ProbeContext
from perf_harness.records import RequestRecord


@dataclass
class DriveState:
    """ArmRun-owned facts that survive scheduler and Judge exceptions."""

    measurement_end_s: float | None = None
    stop: ArmStop = field(default_factory=lambda: ArmStop(reason="aborted"))


async def drive(
    runner: Runner,
    judge: Judge,
    context: ArmContext,
    ctx: ProbeContext,
    cases: list[Case],
    weights: list[float],
    execution: ArmRun,
    state: DriveState,
) -> None:
    load = context.load
    rng = random.Random(load.seed)
    arrival_rng = random.Random(load.seed)
    active: set[asyncio.Task] = set()
    wake = asyncio.Event()
    failures: list[BaseException] = []
    completed = errors = 0
    volume = 0.0
    due = 0.0 if load.saturated else load.arrival_time(volume)
    reason, snapshot = "deadline", None

    def now() -> float:
        return time.monotonic() - ctx.t0

    async def fire(record: RequestRecord, case: Case) -> None:
        nonlocal completed, errors
        record.dispatched_at = now()
        record.state = "dispatched"
        record.operation_run_id = record.id
        operation = Operation(name=runner.name)
        try:
            operation = runner.operation(FireContext(context, case))
            outcome = await runner.fire(FireContext(context, case))
        except asyncio.CancelledError:
            record.state = "interrupted"
            record.reason = "cancelled"
            outcome = Outcome(
                status=None,
                duration_ms=(now() - record.dispatched_at) * 1000,
                meta={"interrupted": True},
                case_id=case.id,
                facets=dict(case.facets),
            )
            raise
        except Exception as error:
            outcome = Outcome(
                status=None,
                duration_ms=(now() - record.dispatched_at) * 1000,
                meta={"exc": type(error).__name__, "exc_detail": str(error)},
            )
        finally:
            record.finished_at = now()
            ctx.stats.done()
            # Even cancellation has an actual-call record, but never a fabricated response.
            if record.state == "interrupted":
                execution.operation_runs.append(
                    OperationRun(
                        id=record.id, service=context.service, operation=operation, outcome=outcome
                    )
                )
        record.state = "finished"
        outcome.case_id = case.id
        outcome.facets = {**case.facets, **outcome.facets}
        record.facets = dict(outcome.facets)
        execution.operation_runs.append(
            OperationRun(
                id=record.id, service=context.service, operation=operation, outcome=outcome
            )
        )
        evaluation = judge(outcome)
        execution.evaluations[record.id] = evaluation
        completed += 1
        errors += not evaluation.ok

    def settled(task: asyncio.Task) -> None:
        active.discard(task)
        if not task.cancelled() and task.exception() is not None:
            failures.append(task.exception())
        wake.set()

    def offer(scheduled: float, drop_reason: str | None = None) -> None:
        case = rng.choices(cases, weights=weights, k=1)[0]
        record = RequestRecord(
            id=f"{execution.id}:{len(execution.requests)}",
            case_id=case.id,
            scheduled_at=scheduled,
            arrived_at=now(),
            facets=dict(case.facets),
        )
        execution.requests.append(record)
        _, cap = load.target(record.arrived_at)
        if drop_reason or ctx.stats.inflight >= cap:
            record.state, record.reason, record.finished_at = (
                "dropped",
                drop_reason or "concurrency_limit",
                record.arrived_at,
            )
            ctx.stats.dropped += 1
            return
        # Reserve before creating a Task: a batch of due arrivals must see previous reservations.
        ctx.stats.start()
        task = asyncio.create_task(fire(record, case))
        active.add(task)
        task.add_done_callback(settled)

    try:
        while now() < load.duration_s:
            if failures:
                raise failures[0]
            if (
                load.abort_on_error_rate is not None
                and completed >= load.breaker_min_n
                and errors / completed >= load.abort_on_error_rate
            ):
                reason = "error_rate"
                snapshot = StopSnapshot(
                    now(), completed, errors, errors / completed, load.abort_on_error_rate
                )
                break
            elapsed = now()
            wake.clear()
            if load.saturated:
                _, cap = load.target(elapsed)
                for _ in range(max(0, cap - ctx.stats.inflight)):
                    offer(elapsed)
            else:
                # Yield per offer so cancellation progresses even after a stall.
                if due <= elapsed:
                    offer(due)
                    volume += arrival_rng.expovariate(1) if load.arrival == "poisson" else 1
                    due = load.arrival_time(volume)
                    await asyncio.sleep(0)
                    continue
            delay = min(0.02, max(0, load.duration_s - now()))
            if not load.saturated:
                delay = min(delay, max(0, due - now()))
            with suppress(TimeoutError):
                await asyncio.wait_for(wake.wait(), timeout=delay)
        # Account for planned arrivals missed by a stalled generator; never send after deadline.
        if reason == "deadline" and not load.saturated:
            while due < load.duration_s:
                offer(due, "scheduler_deadline")
                volume += arrival_rng.expovariate(1) if load.arrival == "poisson" else 1
                due = load.arrival_time(volume)
                await asyncio.sleep(0)
        state.measurement_end_s = load.duration_s if reason == "deadline" else snapshot.at_s
        state.stop = ArmStop(reason=reason, snapshot=snapshot, inflight_at_stop=ctx.stats.inflight)
        if active:
            await asyncio.wait(active, timeout=load.drain_timeout_s)
    finally:
        # Freeze the boundary before cancellation/join; a failing Judge in drain must
        # neither extend measurement nor erase the completed calls and stop census.
        if state.measurement_end_s is None:
            state.measurement_end_s = min(now(), load.duration_s)
            state.stop = ArmStop(reason="aborted", inflight_at_stop=ctx.stats.inflight)
        # Join cancellations before returning ownership of clients or environment resources.
        remaining = list(active)
        for task in remaining:
            task.cancel()
        if remaining:
            await asyncio.gather(*remaining, return_exceptions=True)
        interrupted = sum(r.state == "interrupted" for r in execution.requests)
        state.stop = replace(state.stop, interrupted=interrupted, force_cancelled=bool(interrupted))
    if failures:
        raise failures[0]

File: sdks/typescript/perf-harness/src/scheduler.ts
import { arrivalTime, saturated, target } from "./load";
import type {
  Arm,
  ArmRun,
  ArmStop,
  Case,
  Outcome,
  RequestRecord,
  StopSnapshot,
} from "./model";
import type { ArmContext, Runner } from "./runner";
import type { Judge } from "./judge";

export async function drive(options: {
  runner: Runner;
  judge: Judge;
  context: ArmContext;
  arm: Arm;
  cases: readonly Case[];
  weights: readonly number[];
  execution: ArmRun;
  signal?: AbortSignal;
}): Promise<{ stop: ArmStop; elapsed_s: number }> {
  const { runner, judge, context, arm, cases, weights, execution, signal } =
      options,
    load = arm.load;
  const start = performance.now(),
    now = () => (performance.now() - start) / 1000;
  const active = new Map<Promise<void>, AbortController>();
  const errors: unknown[] = [];
  let completed = 0,
    failed = 0,
    volume = 0,
    due = saturated(load) ? 0 : arrivalTime(load, 0);
  let reason: ArmStop["reason"] = "deadline",
    snapshot: StopSnapshot | undefined;
  let rngState = (load.seed ?? 0) >>> 0;
  const random = () => {
    rngState = (Math.imul(1664525, rngState) + 1013904223) >>> 0;
    return (rngState + 0.5) / 4294967296;
  };
  const pick = () => {
    let value = random() * weights.reduce((a, b) => a + b, 0);
    for (let i = 0; i < cases.length; i++) {
      value -= weights[i]!;
      if (value < 0) return cases[i]!;
    }
    return cases[cases.length - 1]!;
  };
  function offer(scheduled: number, dropReason?: string): void {
    const item = pick();
    const record: RequestRecord = {
      id: `${execution.id}:${execution.requests.length}`,
      case_id: item.id,
      scheduled_at: scheduled,
      arrived_at: now(),
      state: "arrived",
      facets: { ...item.facets },
    };
    execution.requests.push(record);
    if (dropReason || active.size >= target(load, now())[1]) {
      record.state = "dropped";
      record.reason = dropReason ?? "concurrency_limit";
      record.finished_at = now();
      return;
    }
    const controller = new AbortController();
    record.dispatched_at = now();
    record.operation_run_id = record.id;
    record.state = "dispatched";
    // Promise callbacks start after the synchronous reservation below.
    const task = Promise.resolve()
      .then(async () => {
        let outcome: Outcome;
        const fireContext = {
          ...context,
          case: item,
          signal: controller.signal,
        };
        let operation = { name: runner.name };
        try {
          operation = runner.operation?.(fireContext) ?? operation;
          outcome = await runner.fire(fireContext);
        } catch (error) {
          outcome = {
            status: null,
            duration_ms: (now() - record.dispatched_at!) * 1000,
            meta: controller.signal.aborted
              ? { interrupted: true }
              : {
                  exc: error instanceof Error ? error.name : "Error",
                  exc_detail: String(error),
                },
          };
        }
        record.finished_at = now();
        record.state = controller.signal.aborted ? "interrupted" : "finished";
        if (record.state === "interrupted") record.reason = "cancelled";
        outcome = {
          ...outcome,
          case_id: item.id,
          facets: { ...item.facets, ...outcome.facets },
        };
        record.facets = outcome.facets!;
        execution.operation_runs.push({
          id: record.id,
          service: context.service,
          operation,
          outcome,
        });
        if (record.state === "finished") {
          const evaluation = judge(outcome);
          execution.evaluations[record.id] = evaluation;
          completed++;
          if (!evaluation.ok) failed++;
        }
      })
      .catch((error) => {
        errors.push(error);
      })
      .finally(() => {
        active.delete(task);
      });
    active.set(task, controller);
  }
  // Resolve on completion or deadline, with no orphaned timers/listeners.
  async function wait(seconds: number): Promise<void> {
    let timer: ReturnType<typeof setTimeout> | undefined;
    let abort: (() => void) | undefined;
    try {
      await Promise.race([
        ...active.keys(),
        new Promise<void>((resolve) => {
          timer = setTimeout(resolve, Math.max(0, seconds) * 1000);
        }),
        new Promise<void>((resolve) => {
          abort = () => resolve();
          signal?.addEventListener("abort", abort, { once: true });
          if (signal?.aborted) resolve();
        }),
      ]);
    } finally {
      if (timer !== undefined) clearTimeout(timer);
      if (abort) signal?.removeEventListener("abort", abort);
    }
  }
  try {
    while (now() < load.duration_s && !signal?.aborted) {
      if (errors.length) throw errors[0];
      if (
        load.abort_on_error_rate !== undefined &&
        completed >= (load.breaker_min_n ?? 20) &&
        failed / completed >= load.abort_on_error_rate
      ) {
        reason = "error_rate";
        snapshot = {
          at_s: now(),
          completed,
          errors: failed,
          error_rate: failed / completed,
          threshold: load.abort_on_error_rate,
        };
        break;
      }
      if (saturated(load)) {
        const count = Math.max(0, target(load, now())[1] - active.size);
        for (let i = 0; i < count; i++) offer(now());
      } else if (due <= now()) {
        offer(due);
        volume += load.arrival === "poisson" ? -Math.log(random()) : 1;
        due = arrivalTime(load, volume);
        await new Promise<void>((resolve) => setImmediate(resolve));
        continue;
      }
      const delay = Math.max(
        0,
        Math.min(
          0.01,
          load.duration_s - now(),
          saturated(load) ? Infinity : due - now(),
        ),
      );
      await wait(delay);
      // Even immediately completed Runners must let timers and external aborts progress.
      await new Promise<void>((resolve) => setImmediate(resolve));
    }
    if (signal?.aborted) reason = "aborted";
    if (reason === "deadline" && !saturated(load)) {
      while (due < load.duration_s && !signal?.aborted) {
        offer(due, "scheduler_deadline");
        volume += load.arrival === "poisson" ? -Math.log(random()) : 1;
        due = arrivalTime(load, volume);
        await new Promise<void>((resolve) => setImmediate(resolve));
      }
    }
    const elapsed_s = reason === "deadline" ? load.duration_s : now();
    const inflight_at_stop = active.size;
    const drainDeadline = now() + (load.drain_timeout_s ?? 30);
    while (!signal?.aborted && active.size && now() < drainDeadline)
      await wait(drainDeadline - now());
    if (signal?.aborted) reason = "aborted";
    for (const controller of active.values()) controller.abort();
    // Runner must cooperate with cancellation; do not close its clients while work remains.
    await Promise.all(active.keys());
    if (errors.length) throw errors[0];
    const interrupted = execution.requests.filter(
      (r) => r.state === "interrupted",
    ).length;
    return {
      stop: {
        reason,
        snapshot,
        inflight_at_stop,
        interrupted,
        force_cancelled: interrupted > 0,
      },
      elapsed_s,
    };
  } finally {
    for (const controller of active.values()) controller.abort();
    await Promise.all(active.keys());
  }
}
